pkg/executor/show.go

```diff
@@ -904,6 +906,89 @@ func (e *ShowExec) fetchShowIndex() error {
 	return nil
 }
 
+// prefetchColumnNDVs prepares a map of colID->NDV using cache-first and storage-fallback without mutating the stats cache
+func (e *ShowExec) prefetchColumnNDVs(tb table.Table, statsTbl *statistics.Table, pkCol *table.Column) map[int64]int64 {
+	ndvByColID := make(map[int64]int64)
+	needed := make(map[int64]struct{})
+	if statsTbl.Pseudo {
+		if pkCol != nil {
+			needed[pkCol.ID] = struct{}{}
+		}
+		for _, idx := range tb.Indices() {
+			idxInfo := idx.Meta()
+			if idxInfo.State != model.StatePublic {
+				continue
+			}
+			for _, ic := range idxInfo.Columns {
+				tblCol := tb.Meta().Columns[ic.Offset]
+				if tblCol.Hidden {
+					continue
+				}
+				needed[tblCol.ID] = struct{}{}
+			}
+		}
+	} else {
+		if pkCol != nil {
+			if colStats := statsTbl.GetCol(pkCol.ID); colStats != nil {
+				ndvByColID[pkCol.ID] = colStats.NDV
+			} else {
+				needed[pkCol.ID] = struct{}{}
+			}
+		}
+		for _, idx := range tb.Indices() {
+			idxInfo := idx.Meta()
+			if idxInfo.State != model.StatePublic {
+				continue
+			}
+			for _, ic := range idxInfo.Columns {
+				tblCol := tb.Meta().Columns[ic.Offset]
+				if tblCol.Hidden {
+					continue
+				}
+				if colStats := statsTbl.GetCol(tblCol.ID); colStats != nil {
+					ndvByColID[tblCol.ID] = colStats.NDV
+				} else {
+					needed[tblCol.ID] = struct{}{}
+				}
+			}
+		}
+	}
+	if len(needed) > 0 {
+		colIDs := make([]string, 0, len(needed))
+		for id := range needed {
+			colIDs = append(colIDs, strconv.FormatInt(id, 10))
+		}
+		if m, err := fetchColumnNDVs(e.Ctx(), tb.Meta().ID, colIDs); err == nil {
+			for k, v := range m {
+				if _, exists := ndvByColID[k]; !exists {
+					ndvByColID[k] = v
+				}
+			}
+		}
+	}
+	return ndvByColID
+}
+
+const selectNDVForColumnsQuery = "SELECT hist_id, distinct_count FROM mysql.stats_histograms WHERE table_id = %? AND is_index = 0 AND hist_id IN (%?)"
+
+// fetchColumnNDVs loads NDVs for the given column IDs from storage in one round-trip without mutating stats cache
+func fetchColumnNDVs(ctx sessionctx.Context, tableID int64, columnIDs []string) (map[int64]int64, error) {
+	ndvByColID := make(map[int64]int64, len(columnIDs))
+	if len(columnIDs) == 0 {
+		return ndvByColID, nil
+	}
+	rows, _, err := statsutil.ExecRows(ctx, selectNDVForColumnsQuery, tableID, columnIDs)
```

Comment:
It seems unsafe to run another query with the same `sessionctx` of the current query and the current query has not finished yet.
Better to grab a new session from the session pool to run this query.

Reply:
thanks, that makes sense!